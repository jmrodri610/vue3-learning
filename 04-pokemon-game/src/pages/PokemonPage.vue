<template>
<div>
  <h1 v-if="!pokemon">Wait...</h1>
  <div v-else>
    <img src="@/assets/pokeball_PNG8.png" alt="pokeball" class="img-pokeball pokeball">
    <h1>Â¿Cual es este pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>
    <div class="fade-in" v-if="showAnswer">
      <h2 v-if="showAnswer">{{showMessage}}</h2>
      <button @click="newGame">Retry</button>
    </div>
  </div>
</div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from '@/helpers/getPokemonOptions';


export default {
  name: "PokemonPage",
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      showMessage: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const choosenPokemon = Math.floor(Math.random()*4)
      this.pokemon = this.pokemonArr[choosenPokemon]
    },
    checkAnswer (pokemonId) {
      this.showPokemon = true

      if(pokemonId === this.pokemon.id) {
        this.showMessage = "Great job!"
      } else {
        this.showMessage = "Nooo! Wrong answer! Try again!"
      }
        this.showAnswer = true
    },
    async newGame() {
      this.showMessage = '';
      this.pokemon = null;
      this.showAnswer = false;
      this.showPokemon = false;
      await this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray();
  }
};
</script>

<style scoped>
  .img-pokeball {
    width: 100px;
    height: 100px;
  }
</style>
