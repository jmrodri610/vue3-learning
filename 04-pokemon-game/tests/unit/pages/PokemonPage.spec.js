import { shallowMount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage component', ()=> {

    let wrapper;

    beforeEach(()=> {

        wrapper = shallowMount(PokemonPage);
    })

    test('should match with the snapshot', ()=> {

        expect(wrapper.html()).toMatchSnapshot();
    })

    test('should call mixPokemonArray on component mount', ()=>{

        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage);

        expect( mixPokemonArraySpy ).toHaveBeenCalled();
    })

    test('should match with the snapshot after the pokemons are loaded', ()=> {
        
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showMessage: '',
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot();
    })

    test('should show the components PokemonPicture and PokemonOptions', ()=> {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showMessage: '',
                }
            }
        })

        expect(wrapper.find('pokemon-picture-stub').exists).toBeTruthy()
        expect(wrapper.find('pokemon-options-stub').exists).toBeTruthy()
        expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe('1')
        expect(wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()
    })

    test('should succeed on correct pokemon selection', async ()=> {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showMessage: '',
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy();
        expect(wrapper.find('h2').text()).toBe('Great job!')
        
    })

    test('should fail on incorrect pokemon selection', async ()=> {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showMessage: '',
                }
            }
        })

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy();
        expect(wrapper.find('h2').text()).toBe('Nooo! Wrong answer! Try again!')

    })
})