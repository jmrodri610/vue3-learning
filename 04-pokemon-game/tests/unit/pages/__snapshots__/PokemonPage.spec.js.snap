// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match with the snapshot 1`] = `
<div>
  <h1>Wait...</h1>
</div>
`;

exports[`PokemonPage component should match with the snapshot after the pokemons are loaded 1`] = `
<div>
  <div><img src="" alt="pokeball" class="img-pokeball pokeball">
    <h1>Â¿Cual es este pokemon?</h1>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
